### ISSUE TITLE
2.4.0 script plugin support ignoring the output of the file copier provider script

### ISSUE THREAD

**Issue by mgsococo** [rundeck/rundeck#1060]
I have a script plugin that provides both file copier and node executor functions.  Since upgrading from 2.2.3 to 2.4.0 (clean install+project import), Rundeck ignores the output filename from the file copier, and attempts to use some internally-generated name.

E.g. when a script is to be run on the remote node from a job, the following happens:

Rundeck calls the file copier provider script, this script performs the copy, and is given a local file named `/opt/rundeck/var/tmp/dispatch1315971509832685444.tmp`, which it copies to the remote node and outputs `/var/rundeck/sw_rundeck_cYM5CGzj` to stdout per the documentation.

When rundeck subsequently calls the execution provider script, it sends the command `chmod +x /var/rundeck/1420058024517-redacted01.sococo.net-dispatch-script.tmp.sh` which is not the name that was output by the copier script.  The file-copy-destination directory is correctly used, but the file name is not.  It also attempts to use the wrong filename when using the shipped SSH execution provider.

The logs don't really indicate anything useful.  My copy script logs that it wrote the file and returned the correct filename, the file is present on the file system.  The rundeck logs indicate the failure when the chmod command is run:

```
text/x-rundeck-log-v2.0^
^2014-12-31T20:33:44Z|stepbegin||{node=rdserver.sococo.net|step=1|stepctx=1|user=mg}|^
^2014-12-31T20:33:44Z|nodebegin||{node=redacted01.sococo.net|step=1|stepctx=1|user=root}|^
^2014-12-31T20:33:45Z|||{node=redacted01.sococo.net|step=1|stepctx=1|user=root}|chmod: cannot access `/var/rundeck/1420058024517-redacted01.sococo.net-dispatch-script.tmp.sh': No such file or directory^
^2014-12-31T20:33:45Z||ERROR|{node=redacted01.sococo.net|step=1|stepctx=1|user=root}|Failed: NonZeroResultCode: Result code: 1^
^2014-12-31T20:33:45Z|nodeend||{executionState=FAILED|failureReason=NonZeroResultCode|node=redacted01.sococo.net|resultCode=1|step=1|stepctx=1|user=root}|Result code: 1^
^2014-12-31T20:33:45Z|stepend||{executionState=FAILED|failedNodes=redacted01.sococo.net|failureReason=NodeDispatchFailure|node=rdserver.sococo.net|step=1|stepctx=1|user=mg}|Node dispatch failed^
^2014-12-31T20:33:45Z||ERROR|{node=rdserver.sococo.net|user=mg}|Execution failed: 1734: [Workflow step failures: {1=Dispatch failed on 1 nodes: [redacted01.sococo.net: NonZeroResultCode: Result code: 1]}, Node failures: {redacted01.sococo.net=[NonZeroResultCode: Result code: 1]}]^
^END^
```

FWIW, my copier logs these lines into syslog when it does the copy procedure:

```
2014-12-31T12:33:44.742060-08:00 rundeck sw_copy #1734[13759]: copying /opt/rundeck/var/tmp/dispatch1315971509832685444.tmp to redacted01.sococo.net:/var/rundeck/
2014-12-31T12:33:45.163107-08:00 rundeck sw_copy #1734[13759]: /opt/rundeck/var/tmp/dispatch1315971509832685444.tmp -> /var/rundeck/sw_rundeck_cYM5CGzj
2014-12-31T12:33:45.228894-08:00 rundeck sw_copy #1734[13759]: Returning to Rundeck: /var/rundeck/sw_rundeck_cYM5CGzj
```

**Comment by thowo** [comment]
Hello,
just hitting this bug, too. Is the 2.4.1 already scheduled for GA? Thanks in advance, Tom

**Comment by gschueler** [comment]
the file copier was modified in 2.4, because it should support a "destination" file copy, for example if the user is using the "Copy File" step, and specifies their own destination path. In that case they intend to copy the file to a specific location.  Also, for script steps, the user can now specify a desired file extension for the copied file, so Rundeck will supply a generated temp file path using the specified extension if present, and tell the file copier to copy to that destination.

**Comment by gschueler** [comment]
for script steps, I guess if the file copier returns a file path (as was the old behavior) it should still use that for executing the script.  I will restore that behavior in the script plugin file copier

**Comment by mgsococo** [comment]
Thanks.

I might mention that if the Rundeck server generates expects the file copier to use a temporary file name, it does not have any knowledge of whether that file exists already or not at the destination.  If the file copier isn't smart it might overwrite something already there.

**Comment by gschueler** [comment]
fix f7818f458a3dd4c3a8e740bbc4195cbb9cacc295

**Comment by gschueler** [comment]
@mgsococo true, rundeck attempts to name the file uniquely based on timestamp and node name so it would not conflict
